src/ringmesh/mesh/surface_mesh.cpp

```diff
@@ -0,0 +1,590 @@
+
+    template < index_t DIMENSION >
+    index_t SurfaceMeshBase< DIMENSION >::polygon_from_vertex_ids(
+        index_t in0, index_t in1 ) const
+    {
+        ringmesh_assert(
+            in0 < this->nb_vertices() && in1 < this->nb_vertices() );
+
+        // Another possible, probably faster, algorithm is to check if the 2
+        // indices
+        // are neighbors in polygons_ and check that they are in the same
+        // polygon
+
+        // Check if the edge is in one of the polygon
+        for( auto poly : range( nb_polygons() ) )
+        {
+            bool found = false;
+            auto prev =
+                polygon_vertex( { poly, nb_polygon_vertices( poly ) - 1 } );
+            for( auto v : range( nb_polygon_vertices( poly ) ) )
+            {
+                auto p = polygon_vertex( { poly, v } );
+                if( ( prev == in0 && p == in1 ) || ( prev == in1 && p == in0 ) )
+                {
+                    found = true;
+                    break;
+                }
+                prev = p;
+            }
+            if( found )
+            {
+                return poly;
+            }
+        }
+        return NO_ID;
+    }
+
+    template < index_t DIMENSION >
+    index_t SurfaceMeshBase< DIMENSION >::vertex_index_in_polygon(
+        index_t polygon_index, index_t vertex_id ) const
+    {
+        ringmesh_assert( polygon_index < nb_polygons() );
+        for( auto v : range( nb_polygon_vertices( polygon_index ) ) )
+        {
+            if( polygon_vertex( { polygon_index, v } ) == vertex_id )
+            {
+                return v;
+            }
+        }
+        return NO_ID;
+    }
+
+    template < index_t DIMENSION >
+    index_t SurfaceMeshBase< DIMENSION >::closest_vertex_in_polygon(
+        index_t polygon_index, const vecn< DIMENSION >& query_point ) const
+    {
+        index_t result{ 0 };
+        double dist{ DBL_MAX };
+        for( auto v_id : range( nb_polygon_vertices( polygon_index ) ) )
+        {
+            double distance =
+                length2( query_point
+                         - this->vertex( polygon_vertex(
+                               ElementLocalVertex( polygon_index, v_id ) ) ) );
+            if( dist > distance )
+            {
+                dist = distance;
+                result = v_id;
+            }
+        }
+        return result;
+    }
+
+    template < index_t DIMENSION >
+    index_t SurfaceMeshBase< DIMENSION >::find_first_polygon_around_vertex(
+        index_t cur_p, index_t vertex_id, index_t& first_polygon ) const
+    {
+        for( auto lv : range( nb_polygon_vertices( cur_p ) ) )
+        {
+            if( polygon_vertex( { cur_p, lv } ) == vertex_id )
+            {
+                first_polygon = cur_p;
+                break;
+            }
+        }
+        return cur_p;
+    }
+
+    template < index_t DIMENSION >
+    std::vector< index_t >
+        SurfaceMeshBase< DIMENSION >::store_polygons_around_vertex(

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 27 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)